File: sdks/sandbox/javascript/src/models/sandboxes.ts
// Copyright 2026 Alibaba Group Holding Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Domain models for sandbox lifecycle.
 *
 * IMPORTANT:
 * - These are NOT OpenAPI-generated types.
 * - They are intentionally stable and JS-friendly.
 *
 * The internal OpenAPI schemas may change frequently; adapters map responses into these models.
 */

export type SandboxId = string;

export interface ImageAuth extends Record<string, unknown> {
  username?: string;
  password?: string;
  token?: string;
}

export interface ImageSpec {
  uri: string;
  auth?: ImageAuth;
}

export interface PlatformSpec extends Record<string, unknown> {
  /**
   * Target operating system for sandbox provisioning.
   */
  os: "linux" | "windows";
  /**
   * Target CPU architecture for sandbox provisioning.
   */
  arch: "amd64" | "arm64";
}

export type ResourceLimits = Record<string, string>;

export type NetworkRuleAction = "allow" | "deny";

export interface NetworkRule extends Record<string, unknown> {
  /**
   * Whether to allow or deny matching targets.
   */
  action: NetworkRuleAction;
  /**
   * FQDN or wildcard domain (e.g., "example.com", "*.example.com").
   * IP/CIDR is not supported in the egress MVP.
   */
  target: string;
}

export interface NetworkPolicy extends Record<string, unknown> {
  /**
   * Default action when no egress rule matches. Defaults to "deny".
   */
  defaultAction?: NetworkRuleAction;
  /**
   * List of egress rules evaluated in order.
   */
  egress?: NetworkRule[];
}

// ============================================================================
// Credential Vault Models
// ============================================================================

export interface CredentialProxyConfig extends Record<string, unknown> {
  /**
   * Enable transparent MITM support required by Credential Vault injection.
   */
  enabled?: boolean;
}

export interface InlineCredentialSource extends Record<string, unknown> {
  /**
   * Credential source type. Defaults to "inline" when omitted.
   */
  type?: "inline";
  /**
   * Write-only inline credential value. This field is accepted in create/patch
   * requests and is never present in Credential Vault state responses.
   */
  value: string;
}

export interface Credential extends Record<string, unknown> {
  /**
   * Sandbox-local credential name.
   */
  name: string;
  /**
   * Write-only credential source.
   */
  source: InlineCredentialSource;
}

export type CredentialMatchScheme = "https" | "http";

export interface CredentialMatch extends Record<string, unknown> {
  /**
   * URL schemes to match. Defaults to HTTPS in the sidecar.
   */
  schemes?: CredentialMatchScheme[];
  /**
   * @deprecated Port is derived from scheme (https→443, http→80). Values other than 80 or 443 are rejected by the server.
   */
  ports?: number[];
  /**
   * Exact FQDNs or leftmost-label wildcards.
   */
  hosts: string[];
  /**
   * HTTP methods to match.
   */
  methods?: string[];
  /**
   * Request paths to match.
   */
  paths?: string[];
}

export type CredentialSubstitutionSurface = "path" | "query" | "header" | "body";

export interface CredentialSubstitution extends Record<string, unknown> {
  /**
   * Name of the sandbox-local credential used as the replacement value.
   */
  credential: string;
  /**
   * Literal placeholder to replace.
   */
  placeholder: string;
  /**
   * Request surfaces where the placeholder may be replaced.
   */
  in: CredentialSubstitutionSurface[];
}

export interface CustomHeaderEntry extends Record<string, unknown> {
  /**
   * Header name to inject.
   */
  name: string;
  /**
   * Name of the sandbox-local credential to inject as this header value.
   */
  credential: string;
}

interface CredentialAuthSubstitutions {
  substitutions?: CredentialSubstitution[];
}

export type CredentialAuth =
  | ({
      type: "bearer";
      credential: string;
    } & CredentialAuthSubstitutions)
  | ({
      type: "basic";
      /**
       * Credential containing pre-encoded base64(username:password).
       */
      credential: string;
    } & CredentialAuthSubstitutions)
  | ({
      type: "apiKey";
      name: string;
      credential: string;
    } & CredentialAuthSubstitutions)
  | ({
      type: "customHeaders";
      headers: CustomHeaderEntry[];
    } & CredentialAuthSubstitutions)
  | ({
      type: "passthrough";
    } & CredentialAuthSubstitutions);

export interface CredentialBinding extends Record<string, unknown> {
  /**
   * Sandbox-local binding name.
   */
  name: string;
  /**
   * Request match for this binding.
   */
  match: CredentialMatch;
  /**
   * Auth injection rule for this binding.
   */
  auth: CredentialAuth;
}

export interface CredentialMetadata {
  name: string;
  /**
   * Source type only; plaintext source material is not returned.
   */
  sourceType: string;
  revision: number;
}

export interface CredentialAuthMetadata {
  type: string;
  /**
   * Public auth parameter name, such as an API key header name.
   */
  name?: string;
}

export interface CredentialBindingMetadata {
  name: string;
  revision: number;
  match?: CredentialMatch;
  /**
   * Sanitized auth metadata. Plaintext credential references and values are not returned.
   */
  auth?: CredentialAuthMetadata;
}

export interface CredentialVaultState {
  revision: number;
  credentials: CredentialMetadata[];
  bindings: CredentialBindingMetadata[];
}

export interface CredentialListResponse {
  revision: number;
  credentials: CredentialMetadata[];
}

export interface CredentialBindingListResponse {
  revision: number;
  bindings: CredentialBindingMetadata[];
}

export interface CredentialMutationSet extends Record<string, unknown> {
  add?: Credential[];
  replace?: Credential[];
  delete?: string[];
}

export interface CredentialBindingMutationSet extends Record<string, unknown> {
  add?: CredentialBinding[];
  replace?: CredentialBinding[];
  delete?: string[];
}

export interface CredentialVaultCreateRequest extends Record<string, unknown> {
  credentials: Credential[];
  bindings: CredentialBinding[];
}

export interface CredentialVaultPatchRequest extends Record<string, unknown> {
  /**
   * Optional optimistic concurrency guard.
   */
  expectedRevision?: number;
  credentials?: CredentialMutationSet;
  bindings?: CredentialBindingMutationSet;
}

// ============================================================================
// Volume Models
// ============================================================================

/**
 * Host path bind mount backend.
 *
 * Maps a directory on the host filesystem into the container.
 * Only available when the runtime supports host mounts.
 */
export interface Host extends Record<string, unknown> {
  /**
   * Absolute path on the host filesystem to mount.
   * Must start with '/' (Unix) or a drive letter such as 'C:\' or 'D:/'
   * (Windows), and be under an allowed prefix.
   */
  path: string;
}

/**
 * Platform-managed named volume backend.
 *
 * Runtime-neutral abstraction for referencing a pre-existing named volume:
 * - Kubernetes: maps to a PersistentVolumeClaim in the same namespace.
 * - Docker: maps to a Docker named volume.
 */
export interface PVC extends Record<string, unknown> {
  /**
   * Name of the platform volume.
   * In Kubernetes this is the PVC name; in Docker this is the named volume name.
   */
  claimName: string;
  /**
   * When true (default), auto-create the volume if it does not exist.
   */
  createIfNotExists?: boolean;
  /**
   * When true, the auto-created volume (Docker named volume or Kubernetes PVC)
   * is removed on sandbox deletion. Pre-existing volumes are never removed.
   */
  deleteOnSandboxTermination?: boolean;
  /**
   * Kubernetes StorageClass name for auto-created PVCs.
   * Null means use cluster default. Ignored for Docker.
   */
  storageClass?: string | null;
  /**
   * Capacity request for auto-created PVCs (e.g. "1Gi").
   * Ignored for Docker.
   */
  storage?: string | null;
  /**
   * Access modes for auto-created PVCs (e.g. ["ReadWriteOnce"]).
   * Ignored for Docker.
   */
  accessModes?: string[] | null;
}

/**
 * Alibaba Cloud OSS mount backend via ossfs.
 *
 * The runtime mounts a host-side OSS path under `storage.ossfs_mount_root`
 * so the container sees the bucket contents at the specified mount path.
 *
 * In Docker runtime, OSSFS backend requires OpenSandbox Server to run on a Linux host with FUSE support.
 */
export interface OSSFS extends Record<string, unknown> {
  /**
   * OSS bucket name.
   */
  bucket: string;
  /**
   * OSS endpoint (e.g., "oss-cn-hangzhou.aliyuncs.com").
   */
  endpoint: string;
  /**
   * ossfs major version used by runtime mount integration.
   * @default "2.0"
   */
  version?: "1.0" | "2.0";
  /**
   * Additional ossfs mount options.
   *
   * - `1.0`: mounts with `ossfs ... -o <option>`
   * - `2.0`: mounts with `ossfs2 mount ... -c <config-file>` and encodes options as `--<option>` lines in the config file
   */
  options?: string[];
  /**
   * OSS access key ID for inline credentials mode.
   */
  accessKeyId: string;
  /**
   * OSS access key secret for inline credentials mode.
   */
  accessKeySecret: string;
}

/**
 * Storage mount definition for a sandbox.
 *
 * Each volume entry contains:
 * - A unique name identifier
 * - Exactly one backend (host, pvc, ossfs) with backend-specific fields
 * - Common mount settings (mountPath, readOnly, subPath)
 */
export interface Volume extends Record<string, unknown> {
  /**
   * Unique identifier for the volume within the sandbox.
   */
  name: string;
  /**
   * Host path bind mount backend (mutually exclusive with pvc, ossfs).
   */
  host?: Host;
  /**
   * Kubernetes PVC mount backend (mutually exclusive with host, ossfs).
   */
  pvc?: PVC;
  /**
   * Alibaba Cloud OSSFS mount backend (mutually exclusive with host, pvc).
   */
  ossfs?: OSSFS;
  /**
   * Absolute path inside the container where the volume is mounted.
   */
  mountPath: string;
  /**
   * If true, the volume is mounted as read-only. Defaults to false (read-write).
   */
  readOnly?: boolean;
  /**
   * Optional subdirectory under the backend path to mount.
   */
  subPath?: string;
}

export type SandboxState =
  | "Creating"
  | "Running"
  | "Pausing"
  | "Paused"
  | "Resuming"
  | "Deleting"
  | "Deleted"
  | "Error"
  | string;

export interface SandboxStatus extends Record<string, unknown> {
  state: SandboxState;
  reason?: string;
  message?: string;
}

/**
 * Current runtime-confirmed Pool allocation for a sandbox.
 */
export interface AllocationSummary extends Record<string, unknown> {
  /**
   * Confirmed allocation mode.
   */
  mode: "pool";
  /**
   * Concrete Pool reference currently allocated to the sandbox.
   */
  poolRef: string;
  /**
   * Current confirmed allocation state.
   */
  state: "allocated";
}

export interface SandboxInfo extends Record<string, unknown> {
  id: SandboxId;
  image?: ImageSpec;
  snapshotId?: string;
  platform?: PlatformSpec;
  entrypoint: string[];
  metadata?: Record<string, string>;
  extensions?: Record<string, string>;
  status: SandboxStatus;
  /**
   * Current runtime-confirmed Pool allocation, when available.
   */
  allocation?: AllocationSummary;
  /**
   * Sandbox creation time.
   */
  createdAt: Date;
  /**
   * Sandbox expiration time (server-side TTL).
   */
  expiresAt: Date | null;
}

export interface LifecycleHook extends Record<string, unknown> {
  /** Command and arguments executed without implicit shell expansion. */
  command: string[];
  /** Maximum execution time in seconds, up to 3 hours (10800 seconds). Defaults to 60. */
  timeoutSeconds?: number;
}

export interface PeriodicLifecycleHook extends Record<string, unknown> {
  /** Name unique among periodic hooks in this sandbox. */
  name: string;
  /** Five-field cron expression or descriptor such as `@hourly` or `@every 30s`. */
  schedule: string;
  /** Command and arguments executed without implicit shell expansion. */
  command: string[];
  /** Maximum execution time in seconds, up to 300. Defaults to 60. */
  timeoutSeconds?: number;
}

/** Extensible container for sandbox lifecycle hooks. */
export interface SandboxLifecycle extends Record<string, unknown> {
  /**
   * Runs after execd is ready and before the user entrypoint on every container
   * start. A failed or timed-out hook prevents the user entrypoint from starting.
   */
  preStart?: LifecycleHook;
  /**
   * Scheduled hooks run while the sandbox is running. Runs of the same named hook
   * never overlap; a scheduled run is skipped while its previous run is active.
   */
  periodic?: PeriodicLifecycleHook[];
}

export interface CreateSandboxRequest extends Record<string, unknown> {
  image?: ImageSpec;
  snapshotId?: string;
  entrypoint?: string[];
  platform?: PlatformSpec;
  /**
   * Whether to require secure access headers for sandbox endpoint access.
   */
  secureAccess?: boolean;
  /**
   * Timeout in seconds (server semantics).
   */
  timeout?: number | null;
  resourceLimits: ResourceLimits;
  resourceRequests?: ResourceLimits;
  env?: Record<string, string>;
  metadata?: Record<string, string>;
  lifecycle?: SandboxLifecycle;
  /**
   * Optional outbound network policy for the sandbox.
   */
  networkPolicy?: NetworkPolicy;
  /**
   * Optional Credential Vault proxy startup settings.
   */
  credentialProxy?: CredentialProxyConfig;
  /**
   * Optional list of volume mounts for persistent storage.
   */
  volumes?: Volume[];
  extensions?: Record<string, unknown>;
}

export interface CreateSandboxFromTemplateRequest extends Record<string, unknown> {
  /**
   * Succeeded fsb template to create the sandbox from.
   *
   * Template mode fixes the workload shape on the server: only metadata,
   * network policy and extensions may accompany the template id, and the
   * timeout is required.
   */
  templateId: string;
  /**
   * Sandbox timeout in seconds (server semantics). Required in template mode.
   */
  timeout: number;
  metadata?: Record<string, string>;
  /**
   * Optional outbound network policy for the sandbox.
   */
  networkPolicy?: NetworkPolicy;
  /**
   * Opaque extension parameters passed through to the server as-is.
   */
  extensions?: Record<string, string>;
}

export interface CreateSandboxResponse extends Record<string, unknown> {
  id: SandboxId;
  status: SandboxStatus;
  platform?: PlatformSpec;
  metadata?: Record<string, string>;
  extensions?: Record<string, string>;
  /**
   * Sandbox expiration time after creation.
   */
  expiresAt: Date | null;
  /**
   * Sandbox creation time.
   */
  createdAt: Date;
  entrypoint: string[];
}

export type SnapshotState = "Creating" | "Deleting" | "Ready" | "Failed" | string;

export interface SnapshotStatus extends Record<string, unknown> {
  state: SnapshotState;
  reason?: string;
  message?: string;
  lastTransitionAt?: Date;
}

export interface SnapshotInfo extends Record<string, unknown> {
  id: string;
  sandboxId: SandboxId;
  name?: string;
  status: SnapshotStatus;
  createdAt: Date;
}

export interface CreateSnapshotRequest extends Record<string, unknown> {
  name?: string;
}

export interface ListSnapshotsResponse extends Record<string, unknown> {
  items: SnapshotInfo[];
  pagination?: PaginationInfo;
}

export interface PaginationInfo extends Record<string, unknown> {
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  hasNextPage: boolean;
}

export interface ListSandboxesResponse extends Record<string, unknown> {
  items: SandboxInfo[];
  pagination?: PaginationInfo;
}

export type SandboxMetadataPatch = Record<string, string | null>;

export interface RenewSandboxExpirationRequest {
  expiresAt: string;
}

export interface RenewSandboxExpirationResponse extends Record<string, unknown> {
  /**
   * Updated expiration time (if the server returns it).
   */
  expiresAt?: Date;
}

export interface Endpoint extends Record<string, unknown> {
  endpoint: string;
  /**
   * Headers that must be included on every request targeting this endpoint
   * (e.g. when the server requires them for routing or auth). Omit or empty if not required.
   */
  headers?: Record<string, string>;
  /**
   * Origin of the sandbox taken from the server's `OPEN-SANDBOX-ORIGIN`
   * response header (see {@link SandboxOrigin}). Omitted when the server
   * does not send it.
   */
  origin?: string;
}

/**
 * Origin backing a sandbox.
 *
 * The protocol defines a single origin value: `template` (reported by the
 * server via the `OPEN-SANDBOX-ORIGIN` response header, and set locally when
 * the sandbox was explicitly created from a template). Anything else -
 * including sandboxes created from an image or a snapshot - is `unknown`.
 *
 * The server may introduce new values in future versions; treat a missing or
 * unknown value as "not template-backed" and keep using the egress sidecar.
 */
export enum SandboxOrigin {
  /**
   * Runs on a fsb golden-image template (no sandbox-side egress sidecar;
   * egress policy goes through the lifecycle control plane).
   */
  TEMPLATE = "template",
  /**
   * The origin could not be determined (create from an image or snapshot,
   * or an older server that does not send the header).
   */
  UNKNOWN = "unknown",
}

export interface ListSandboxesParams {
  /**
   * Filter by lifecycle state (the API supports multiple `state` query params).
   * Example: `{ states: ["Running", "Paused"] }`
   */
  states?: string[];
  /**
   * Filter by metadata key-value pairs.
   * NOTE: This will be encoded to a single `metadata` query parameter as described in the spec.
   */
  metadata?: Record<string, string>;
  page?: number;
  pageSize?: number;
}

export interface ListSnapshotsParams {
  sandboxId?: SandboxId;
  name?: string;
  states?: string[];
  page?: number;
  pageSize?: number;
}

File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/domain/services/Sandboxes.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.domain.services

import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CreateTemplateRequest
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialProxyConfig
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.NetworkPolicy
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PagedSandboxInfos
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PagedSnapshotInfos
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PagedTemplateInfos
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.PlatformSpec
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxCreateResponse
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxEndpoint
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxFilter
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxImageSpec
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxInfo
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxLifecycle
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SandboxRenewResponse
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SnapshotFilter
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.SnapshotInfo
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.TemplateFilter
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.TemplateInfo
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.Volume
import java.time.Duration
import java.time.OffsetDateTime

/**
 * Core sandbox lifecycle management service.
 *
 * This service provides a clean abstraction over sandbox creation, management,
 * and termination operations, completely isolating business logic from API implementation details.
 */
interface Sandboxes {
    /**
     * Creates a new sandbox with the specified configuration.
     *
     * @param spec Container image specification for provisioning the sandbox
     * @param entrypoint The command to run as the sandbox's main process (e.g. `["python", "/app/main.py"]`)
     * @param env Environment variables injected into the sandbox runtime
     * @param metadata User-defined metadata used for management and filtering
     * @param timeout Sandbox lifetime. Pass null to require explicit cleanup.
     * @param resource Runtime resource limits (e.g. cpu/memory). Exact semantics are server-defined
     * @param platform Optional runtime platform constraint used for provisioning
     * @param networkPolicy Optional outbound network policy (egress)
     * @param secureAccess Whether to enable secured access for sandbox endpoints
     * @param extensions Opaque extension parameters passed through to the server as-is. Prefer namespaced keys
     * @param volumes Optional list of volume mounts for persistent storage
     * @param snapshotId Optional snapshot identifier used to restore a sandbox instead of booting from an image
     * @return Sandbox creation response containing the sandbox id
     */
    fun createSandbox(
        spec: SandboxImageSpec?,
        entrypoint: List<String>?,
        env: Map<String, String>,
        metadata: Map<String, String>,
        timeout: Duration?,
        resource: Map<String, String>,
        networkPolicy: NetworkPolicy?,
        extensions: Map<String, String>,
        volumes: List<Volume>?,
        platform: PlatformSpec? = null,
        secureAccess: Boolean = false,
        snapshotId: String? = null,
        resourceRequests: Map<String, String>? = null,
    ): SandboxCreateResponse

    /**
     * Creates a sandbox with optional Credential Vault proxy startup settings.
     */
    fun createSandbox(
        spec: SandboxImageSpec?,
        entrypoint: List<String>?,
        env: Map<String, String>,
        metadata: Map<String, String>,
        timeout: Duration?,
        resource: Map<String, String>,
        networkPolicy: NetworkPolicy?,
        extensions: Map<String, String>,
        volumes: List<Volume>?,
        platform: PlatformSpec? = null,
        secureAccess: Boolean = false,
        snapshotId: String? = null,
        credentialProxy: CredentialProxyConfig?,
        resourceRequests: Map<String, String>? = null,
    ): SandboxCreateResponse {
        if (credentialProxy == null) {
            return createSandbox(
                spec = spec,
                entrypoint = entrypoint,
                env = env,
                metadata = metadata,
                timeout = timeout,
                resource = resource,
                networkPolicy = networkPolicy,
                extensions = extensions,
                volumes = volumes,
                platform = platform,
                secureAccess = secureAccess,
                snapshotId = snapshotId,
                resourceRequests = resourceRequests,
            )
        }
        throw UnsupportedOperationException(
            "Credential Vault proxy is not supported by this Sandboxes implementation",
        )
    }

    /**
     * Creates a sandbox with optional Credential Vault proxy and lifecycle hooks.
     *
     * Existing implementations remain compatible when [lifecycle] is null or empty.
     *
     * @param lifecycle Optional hooks. A value without pre-start or periodic hooks is ignored.
     * @throws UnsupportedOperationException if non-empty hooks are requested from an
     * implementation that does not override this method.
     */
    fun createSandbox(
        spec: SandboxImageSpec?,
        entrypoint: List<String>?,
        env: Map<String, String>,
        metadata: Map<String, String>,
        timeout: Duration?,
        resource: Map<String, String>,
        networkPolicy: NetworkPolicy?,
        extensions: Map<String, String>,
        volumes: List<Volume>?,
        platform: PlatformSpec? = null,
        secureAccess: Boolean = false,
        snapshotId: String? = null,
        credentialProxy: CredentialProxyConfig?,
        resourceRequests: Map<String, String>? = null,
        lifecycle: SandboxLifecycle?,
    ): SandboxCreateResponse {
        if (lifecycle == null || lifecycle.isEmpty) {
            return createSandbox(
                spec = spec,
                entrypoint = entrypoint,
                env = env,
                metadata = metadata,
                timeout = timeout,
                resource = resource,
                networkPolicy = networkPolicy,
                extensions = extensions,
                volumes = volumes,
                platform = platform,
                secureAccess = secureAccess,
                snapshotId = snapshotId,
                credentialProxy = credentialProxy,
                resourceRequests = resourceRequests,
            )
        }
        throw UnsupportedOperationException(
            "Sandbox lifecycle hooks are not supported by this Sandboxes implementation",
        )
    }

    /**
     * Creates a sandbox from a fsb golden-image template.
     *
     * Template mode fixes the workload shape on the server: only [metadata],
     * [networkPolicy] and [extensions] may accompany the [templateId], and the
     * [timeout] is required.
     *
     * @param templateId Unique identifier of a `Succeeded` template owned by the requester
     * @param timeout Sandbox lifetime. Required in template mode.
     * @param metadata User-defined metadata used for management and filtering
     * @param networkPolicy Optional outbound network policy (egress)
     * @param extensions Opaque extension parameters passed through to the server as-is. Prefer namespaced keys
     * @return Sandbox creation response containing the sandbox id
     * @throws UnsupportedOperationException if requested from an implementation that does not override this method
     */
    fun createSandboxFromTemplate(
        templateId: String,
        timeout: Duration,
        metadata: Map<String, String> = emptyMap(),
        networkPolicy: NetworkPolicy? = null,
        extensions: Map<String, String> = emptyMap(),
    ): SandboxCreateResponse =
        throw UnsupportedOperationException(
            "Template-based sandbox creation is not supported by this Sandboxes implementation",
        )

    /**
     * Creates a new fsb golden-image template.
     *
     * The build is asynchronous: the response starts at [TemplatePhase.PENDING];
     * poll [getTemplate] until the status reaches `Succeeded` or `Failed`. Only a
     * `Succeeded` template can create sandboxes.
     *
     * @param request Template build request
     * @return Current template information
     * @throws UnsupportedOperationException if requested from an implementation that does not override this method
     */
    fun createTemplate(request: CreateTemplateRequest): TemplateInfo =
        throw UnsupportedOperationException(
            "Template management is not supported by this Sandboxes implementation",
        )

    /**
     * Retrieves information about an existing template.
     *
     * @param templateId Unique identifier of the template
     * @return Current template information
     * @throws UnsupportedOperationException if requested from an implementation that does not override this method
     */
    fun getTemplate(templateId: String): TemplateInfo =
        throw UnsupportedOperationException(
            "Template management is not supported by this Sandboxes implementation",
        )

    /**
     * Lists templates with optional filtering.
     *
     * @param filter Optional filter criteria
     * @return List of template information matching the filter
     * @throws UnsupportedOperationException if requested from an implementation that does not override this method
     */
    fun listTemplates(filter: TemplateFilter): PagedTemplateInfos =
        throw UnsupportedOperationException(
            "Template management is not supported by this Sandboxes implementation",
        )

    /**
     * Deletes a template by id.
     *
     * Sandboxes already created from the template are unaffected.
     *
     * @param templateId Unique identifier of the template
     * @throws UnsupportedOperationException if requested from an implementation that does not override this method
     */
    fun deleteTemplate(templateId: String) {
        throw UnsupportedOperationException(
            "Template management is not supported by this Sandboxes implementation",
        )
    }

    /**
     * Retrieves information about an existing sandbox.
     *
     * @param sandboxId Unique identifier of the sandbox
     * @return Current sandbox information
     */
    fun getSandboxInfo(sandboxId: String): SandboxInfo

    /**
     * Lists sandboxes with optional filtering.
     *
     * @param filter Optional filter criteria
     * @return List of sandbox information matching the filter
     */
    fun listSandboxes(filter: SandboxFilter): PagedSandboxInfos

    /**
     * Patches sandbox metadata.
     *
     * @param sandboxId Unique identifier of the sandbox
     * @param patch Metadata merge patch. Non-null values add or replace keys; null values delete keys
     * @return Current sandbox information after applying the patch
     */
    fun patchSandboxMetadata(
        sandboxId: String,
        patch: Map<String, String?>,
    ): SandboxInfo

    fun createSnapshot(
        sandboxId: String,
        name: String? = null,
    ): SnapshotInfo

    fun getSnapshot(snapshotId: String): SnapshotInfo

    fun listSnapshots(filter: SnapshotFilter): PagedSnapshotInfos

    fun deleteSnapshot(snapshotId: String)

    /**
     * Get sandbox endpoint
     *
     * @param sandboxId sandbox id
     * @param port endpoint port number
     * @return Target sandbox endpoint
     */
    fun getSandboxEndpoint(
        sandboxId: String,
        port: Int,
    ): SandboxEndpoint

    /**
     * Get sandbox endpoint
     *
     * @param sandboxId sandbox id
     * @param port endpoint port number
     * @param useServerProxy whether to use server proxy for endpoint (default false)
     * @return Target sandbox endpoint
     */
    fun getSandboxEndpoint(
        sandboxId: String,
        port: Int,
        useServerProxy: Boolean,
    ): SandboxEndpoint

    /**
     * Get signed sandbox endpoint with an OSEP-0011 route token.
     *
     * @param sandboxId sandbox id
     * @param port endpoint port number
     * @param expires Unix epoch seconds for the signed route token expiry
     * @param useServerProxy whether to use server proxy for endpoint (default false)
     * @return Target sandbox endpoint
     */
    fun getSignedSandboxEndpoint(
        sandboxId: String,
        port: Int,
        expires: Long,
        useServerProxy: Boolean = false,
    ): SandboxEndpoint

    /**
     * Pauses a running sandbox, preserving its state.
     *
     * @param sandboxId Unique identifier of the sandbox
     */
    fun pauseSandbox(sandboxId: String)

    /**
     * Resumes a paused sandbox.
     *
     * @param sandboxId Unique identifier of the sandbox
     */
    fun resumeSandbox(sandboxId: String)

    /**
     * Renew the expiration time of a sandbox.
     *
     * @param sandboxId Unique identifier of the sandbox
     * @param newExpirationTime New expiration timestamp
     *
     * @return Sandbox renew response with new expire info
     */
    fun renewSandboxExpiration(
        sandboxId: String,
        newExpirationTime: OffsetDateTime,
    ): SandboxRenewResponse

    /**
     * Terminates a sandbox and releases all associated resources.
     *
     * @param sandboxId Unique identifier of the sandbox
     */
    fun killSandbox(sandboxId: String)

    /** Remove all cached endpoints for a sandbox. No-op if caching is disabled. */
    fun invalidateEndpointCache(sandboxId: String) {}
}
